typing/ctype.ml

```diff
@@ -5450,17 +5450,47 @@ let same_constr env t1 t2 =
 let () =
   Env.same_constr := same_constr
 
-let immediacy env typ =
-   match get_desc typ with
-  | Tconstr(p, _args, _abbrev) ->
-    begin try
-      let type_decl = Env.find_type p env in
-      type_decl.type_immediate
-    with Not_found -> Type_immediacy.Unknown
-    (* This can happen due to e.g. missing -I options,
-       causing some .cmi files to be unavailable.
-       Maybe we should emit a warning. *)
+(* We use expand_head_opt version of expand_head to get access
+   to the manifest type of private abbreviations. *)
+let rec get_unboxed_type_representation env ty fuel =
+  if fuel < 0 then ty else
+  let ty = expand_head_opt env ty in
+  match get_desc ty with
+  | Tconstr (p, args, _) ->
+    begin match Env.find_type p env with
+    | exception Not_found -> ty
+    | {type_params; type_kind =
+         Type_record ([{ld_type = ty2; _}], Record_unboxed _)
+       | Type_variant ([{cd_args = Cstr_tuple [ty2]; _}], Variant_unboxed)
+       | Type_variant ([{cd_args = Cstr_record [{ld_type = ty2; _}]; _}],
+                       Variant_unboxed)}
+      ->
+        let ty2 = match get_desc ty2 with Tpoly (t, _) -> t | _ -> ty2 in
+        get_unboxed_type_representation env
+          (apply env type_params ty2 args) (fuel - 1)
+    | _ -> ty
     end
+  | _ -> ty
+
+let get_unboxed_type_representation env ty =
+  (* Do not give too much fuel: PR#7424 *)
+  get_unboxed_type_representation env ty 100
+
+let kind_immediacy = function
+  | Type_open | Type_record _ -> Type_immediacy.Unknown
+  | Type_abstract { immediate } -> immediate
+  | Type_variant (cstrs, _) ->
+     if List.exists (fun c -> c.Types.cd_args <> Types.Cstr_tuple []) cstrs
+     then Type_immediacy.Unknown
+     else Type_immediacy.Always
```

Comment:
I agree it would great to have a way to record that types are never immediate.

In addition to clarifying this code, it could be a performance win - as you observed below, the immediacy recorded in type kinds is an upper bound, so when we see `Type_immediacy.Unknown` on a type abbreviation we sometimes expand the manifest to check whether the upper bound was conservative.  This is potentially expensive, and could be skipped if we had a way to record types were `NonImmediate` even after substitutions.  We've been considering different designs for this in the context of our "unboxed types" prototype (which essentially just expands the lattice you suggest with still more options - different "layouts"), but haven't settled on anything and plan to revisit it in the future.